File: internal/protocol/semantic_types.go
package protocol

type ImmediateByte uint8
type StringID uint32

const StringIDNone StringID = 0

func StringIDFromStored(raw int64) (StringID, bool) {
	if raw > 0 || raw == -1<<63 {
		return 0, false
	}
	if raw == 0 {
		return StringIDNone, true
	}
	value := uint64(-raw)
	if value > 1<<32-1 {
		return 0, false
	}
	return StringID(value), true
}

func (id StringID) ToStored() int64 {
	if id == 0 {
		return 0
	}
	return -int64(id)
}

type ProcessID int32

const InvalidProcessID ProcessID = -1

type ProcessStatus struct{ Raw int32 }

func (s ProcessStatus) Exited() int32 { return (s.Raw >> 8) & 0xff }

type SourceID uint32
type FileID uint32
type ModuleID uint32

const (
	SourceIDNone SourceID = 0
	FileIDNone   FileID   = 0
	ModuleIDNone ModuleID = 0
)

type Count uint64
type Index uint64
type SourceOffset uint64
type RawDumpWord struct{ Bits int64 }

File: src/graph/semantic_types.zig
//! Language-neutral semantic handle vocabulary.
//!
//! These types are the authoritative mapping used by the automated Go port.
//! Their integer representation is deliberately explicit, but distinct enum
//! types prevent accidental interchange inside the implementation.

const std = @import("std");

pub const ImmediateByte = enum(u8) {
    _,
};

pub const StringID = enum(u32) {
    none = 0,
    _,

    pub fn fromStored(raw: i64) ?StringID {
        if (raw >= 0) return if (raw == 0) .none else null;
        return @enumFromInt(@as(u32, @intCast(-raw)));
    }

    pub fn toStored(self: StringID) i64 {
        const raw: i64 = @intCast(@intFromEnum(self));
        return if (raw == 0) 0 else -raw;
    }
};

pub const ProcessID = enum(i32) {
    invalid = -1,
    _,
};

pub const ProcessStatus = packed struct {
    raw: i32,

    pub fn exited(self: ProcessStatus) i32 {
        return (self.raw >> 8) & 0xff;
    }
};

pub const SourceID = enum(u32) {
    none = 0,
    _,
};

pub const FileID = enum(u32) {
    none = 0,
    _,
};

pub const ModuleID = enum(u32) {
    none = 0,
    _,
};

pub const Count = enum(u64) {
    _,
};

pub const Index = enum(u64) {
    _,
};

pub const SourceOffset = enum(u64) {
    _,
};

/// Exact signed word stored by the `.x` codec. It must never be interpreted
/// without an explicit typed constructor at the codec boundary.
pub const RawDumpWord = packed struct {
    bits: i64,
};

test "semantic IDs cannot be implicitly interchanged and preserve wire values" {
    const string_id: StringID = @enumFromInt(7);
    try std.testing.expectEqual(@as(i64, -7), string_id.toStored());
    try std.testing.expectEqual(string_id, StringID.fromStored(-7).?);
    try std.testing.expectEqual(StringID.none, StringID.fromStored(0).?);
    try std.testing.expect(StringID.fromStored(7) == null);
    try std.testing.expect(@TypeOf(@as(FileID, .none)) != @TypeOf(@as(ModuleID, .none)));
}
